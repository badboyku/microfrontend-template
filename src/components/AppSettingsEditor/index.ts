/* istanbul ignore file */
import AppSettingsEditor from './AppSettingsEditor';

export default AppSettingsEditor;
