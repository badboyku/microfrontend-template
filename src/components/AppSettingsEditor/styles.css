.settings-button {
  background: none;
  border: none;
  float: right;
  position: absolute;
  outline: 0;
  top: 0;
  right: 0;
  height: 16px;
  width: 16px;
  z-index: 5000;
}

.settings-container {
  background-color: rgba(15, 15, 15, 0.3);
  position: fixed;
  inset: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100vw;
  height: 100vh;
  z-index: 10000;
}

.settings-modal {
  background-color: rgba(255, 255, 255, 0.95);
  box-shadow:
    whitesmoke 1px 2px,
    white 1px 1px;
  border-radius: 10px;
  padding: 10px 15px;
  width: 50vw;
  min-width: 485px;
}

/* Header */
.settings-modal-header-row {
  display: flex;
  flex-direction: row;
}
.settings-modal-header {
  width: 100%;
}
.settings-modal-title {
  color: black;
  font-size: 24px;
  font-weight: bold;
}
.settings-modal-header-close {
  transform: translate(10px, -15px);
}
.settings-modal-header-close-button {
  color: grey;
  cursor: pointer;
  font-size: 40px;
  font-weight: bold;
}
.settings-modal-header-close-button:hover {
  color: black;
}

/* Form */
.settings-form-field-row {
  display: flex;
  align-items: baseline;
  padding-bottom: 5px;
}
.settings-form-field-label {
  color: black;
  cursor: pointer;
  font-size: 14px;
  padding-right: 5px;
  width: 132px;
  min-width: 132px;
  max-width: 132px;
  white-space: nowrap;
}
.settings-form-field-input {
  width: 100%;
}
.settings-form-field-input-box {
  background-color: white;
  border: 1px solid grey;
  color: black;
  font-size: 14px;
  width: 98%;
  padding: 4px;
}
.settings-form-field-help {
  font-size: 12px;
}
.settings-form-button-row {
  display: flex;
  gap: 10px;
  margin-left: 135px;
  margin-right: 10px;
  padding-top: 10px;
  padding-bottom: 10px;
}
.settings-form-button {
  background-color: #efefef;
  border: 1px solid grey;
  border-radius: 5px;
  color: black;
  cursor: pointer;
  font-size: 16px;
  font-weight: bold;
  padding: 2px 10px;
}
.settings-form-button:hover {
  background-color: #dedede;
}
