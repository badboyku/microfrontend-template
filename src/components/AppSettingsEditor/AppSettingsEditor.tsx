import { useCallback, useEffect, useState } from 'react';
import { useLocation, useNavigate, useRevalidator } from 'react-router-dom';
import logger from 'utils/logger';
import settings from 'utils/settings';
import type { AppSettings } from '@types';
import type { ChangeEvent, FormEvent, KeyboardEvent, MouseEvent } from 'react';
import './styles.css';

const fields: { name: keyof AppSettings; label: string; help?: string }[] = [{ name: 'token', label: 'Token' }];

type Props = {};

const AppSettingsEditor = (_props: Props) => {
  logger.debug('microfrontend-template: components/AppSettingsEditor called');
  const { pathname, search } = useLocation();
  const navigate = useNavigate();
  const { revalidate } = useRevalidator();

  const [isModalOpen, setIsModalOpen] = useState(false);
  const [isMouseLeftButtonDown, setIsMouseLeftButtonDown] = useState(false);
  const [keys, setKeys] = useState('');
  const [settingsState, setSettingsState] = useState<AppSettings>(settings.getSettings());

  const { token } = settingsState;

  useEffect(() => {
    /* istanbul ignore next */
    if (keys === SETTINGS_CODE) {
      setIsModalOpen(true);
    }
  }, [keys]);

  // Handle container events.
  /* istanbul ignore next */
  const containerHandleOnKeyDown = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
    event.stopPropagation();
    if (event.key === 'Escape') {
      setIsModalOpen(false);
    }
  }, []);
  /* istanbul ignore next */
  const containerHandleOnMouseDown = useCallback((event: MouseEvent<HTMLInputElement>) => {
    event.stopPropagation();
    const element = event.target as HTMLDivElement;
    if (element?.getAttribute('role') === 'presentation') {
      setIsModalOpen(false);
    }
  }, []);

  // Handle close button events.
  /* istanbul ignore next */
  const closeButtonHandleOnKeyDown = useCallback((event: KeyboardEvent<HTMLInputElement>) => {
    event.stopPropagation();
    if ([' ', 'Enter', 'Escape'].includes(event.key)) {
      setIsModalOpen(false);
    }
  }, []);
  /* istanbul ignore next */
  const closeButtonHandleOnMouseDown = useCallback((event: MouseEvent<HTMLInputElement>) => {
    event.stopPropagation();
    setIsModalOpen(false);
  }, []);

  // Handle form events.
  /* istanbul ignore next */
  const formHandleSubmit = useCallback(
    (event: FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      const newSettings = {
        ...(settings.token !== token ? { auth: { isAuthorized: false, authorizedDateTime: undefined }, token } : {}),
      };
      settings.updateSettings(newSettings, true);
      setIsModalOpen(false);

      const searchParams = new URLSearchParams(search);
      const params = searchParams.toString();
      const queryParams = params ? `?${params}` : '';
      navigate(`${pathname}${queryParams}`);

      revalidate();
    },
    [token, search, navigate, pathname, revalidate],
  );
  /* istanbul ignore next */
  const formInputHandleChange = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    const { name, value } = event.target;
    setSettingsState((prevState) => ({ ...prevState, [name]: value }));
  }, []);

  // Handle settings button events.
  /* istanbul ignore next */
  const settingsButtonHandleOnKeyDown = useCallback(
    (event: KeyboardEvent<HTMLButtonElement>) => {
      event.stopPropagation();
      if (isMouseLeftButtonDown && !isModalOpen) {
        setKeys(`${keys}${event.key}`);
      }
    },
    [isMouseLeftButtonDown, isModalOpen, keys],
  );
  /* istanbul ignore next */
  const settingsButtonHandleOnMouseDown = useCallback(
    (event: MouseEvent<HTMLButtonElement>) => {
      event.stopPropagation();
      if (!isModalOpen && event.button === 0) {
        setIsMouseLeftButtonDown(true);
      }
    },
    [isModalOpen],
  );
  /* istanbul ignore next */
  const settingsButtonHandleOnMouseUp = useCallback((event: MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    setIsMouseLeftButtonDown(false);
    setKeys('');
  }, []);

  /* istanbul ignore next */
  if (isModalOpen) {
    return (
      <div
        role="presentation"
        className="settings-container"
        onKeyDown={containerHandleOnKeyDown}
        onMouseDown={containerHandleOnMouseDown}
      >
        <div className="settings-modal">
          <div className="settings-modal-header-row">
            <div className="settings-modal-header">
              <div className="settings-modal-title">App Settings (microfrontend-template)</div>
            </div>
            <div className="settings-modal-header-close">
              <span
                role="button"
                className="settings-modal-header-close-button"
                onKeyDown={closeButtonHandleOnKeyDown}
                onMouseDown={closeButtonHandleOnMouseDown}
                tabIndex={0}
              >
                &#9746;
              </span>
            </div>
          </div>
          <form autoComplete="off" onSubmit={formHandleSubmit}>
            {fields.map(({ name, label, help }) => {
              const value = settingsState[name];

              return (
                <div key={name}>
                  <label htmlFor={name}>
                    <div className="settings-form-field-row">
                      <div className="settings-form-field-label">{label}:</div>
                      <div className="settings-form-field-input">
                        <input
                          className="settings-form-field-input-box"
                          type="text"
                          id={name}
                          name={name}
                          value={typeof value === 'string' || typeof value === 'boolean' ? value.toString() : ''}
                          onChange={formInputHandleChange}
                        />
                        {help && <span className="settings-form-field-help">{help}</span>}
                      </div>
                    </div>
                  </label>
                </div>
              );
            })}
            <div className="settings-form-button-row">
              <div>
                <input className="settings-form-button" type="submit" value="Save" />
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }

  return (
    <button
      id="settings-button"
      className="settings-button"
      type="button"
      aria-hidden
      onKeyDown={settingsButtonHandleOnKeyDown}
      onMouseDown={settingsButtonHandleOnMouseDown}
      onMouseUp={settingsButtonHandleOnMouseUp}
      tabIndex={0}
    />
  );
};

export default AppSettingsEditor;
