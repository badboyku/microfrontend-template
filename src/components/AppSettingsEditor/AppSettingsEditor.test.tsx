import { render } from '@testing-library/react';
import { useLocation, useNavigate, useRevalidator } from 'react-router-dom';
import AppSettingsEditor from './AppSettingsEditor';

jest.mock('react-router-dom');
jest.mock('utils/logger');

const renderAppSettingsEditor = (props = {}) => {
  return render(<AppSettingsEditor {...props} />);
};

describe('Component AppSettingsEditor', () => {
  const location = { pathname: '/foo', search: '?foo=bar' };
  const navigate = jest.fn();
  const revalidate = jest.fn();

  it('renders without crashing', () => {
    jest.mocked(useLocation).mockReturnValue(location as never);
    jest.mocked(useNavigate).mockReturnValue(navigate);
    jest.mocked(useRevalidator).mockReturnValue({ revalidate } as never);

    const { asFragment } = renderAppSettingsEditor();

    expect(asFragment()).toMatchSnapshot();
  });
});
