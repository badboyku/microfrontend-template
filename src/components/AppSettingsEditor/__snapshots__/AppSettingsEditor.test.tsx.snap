// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Component AppSettingsEditor renders without crashing 1`] = `
<DocumentFragment>
  <button
    aria-hidden="true"
    class="settings-button"
    id="settings-button"
    tabindex="0"
    type="button"
  />
</DocumentFragment>
`;
